### ISSUE TITLE
Elixir 1.3 compiler intermittently crashing when compiling dependencies on my project

### ISSUE THREAD

**Issue by myronmarston** [elixir-lang/elixir#4772]
### Environment

* Elixir version (elixir -v):

```
Erlang/OTP 18 [erts-7.2.1] [source] [64-bit] [smp:4:4] [async-threads:10] [hipe] [kernel-poll:false] [dtrace]

Elixir 1.3.0-rc.0 (7881123)
```
* Operating system:

OS X el capitan

### Current behavior

I've been trying to get our project on Elixir 1.3.0-rc.0.  Twice, the compiler has crashed in a non-reproducible way.  In both cases, it appears to be related to struct definitions.

The first time I encountered this error while it was compiling timex 1.0.2:

```
== Compilation error on file lib/parse/datetime/parser.ex ==
** (UndefinedFunctionError) function Timex.DateTime.__struct__/0 is undefined or private. Did you mean one of:

      * __struct__/0
      * __struct__/1

    Timex.DateTime.__struct__()
    (elixir) Kernel.Typespec.typespec/3
    (elixir) lib/kernel/typespec.ex:716: anonymous fn/4 in Kernel.Typespec.typespec/3
    (elixir) lib/enum.ex:1557: Enum."-reduce/3-lists^foldl/2-0-"/3
    (elixir) lib/kernel/typespec.ex:716: Kernel.Typespec.typespec/3
    (elixir) lib/kernel/typespec.ex:996: Kernel.Typespec.fn_args/5
    (elixir) lib/kernel/typespec.ex:460: Kernel.Typespec.translate_spec/7
```

Another time I got this while it was compiling ExVCR 0.7.2:

```
== Compilation error on file lib/exvcr/adapter/ibrowse/converter.ex ==
** (CompileError) lib/exvcr/adapter/ibrowse/converter.ex:48: ExVCR.Response.__struct__/1 is undefined, cannot expand struct ExVCR.Response
    lib/exvcr/adapter/ibrowse/converter.ex:47: (module)
```

Each time, when I've tried again (or even blown away `_build` so I can try again from scratch), I have not been able to reproduce -- so there appears to be some source of non-determinism at play here.

### Expected behavior

I expect the elixir compile to consistently produce the same behavior if I do `rm -rf _build && mix compile`.  If there's a compile error I need to address, it should fail consistently.  If there's not, it shouldn't fail at all.

**Comment by josevalim** [comment]
I believe this is the same error @Gazler is seeing.

**Comment by Gazler** [comment]
The errors look very similar. Mine have not been failing on dependencies, but I am using an umbrella application. I have been trying to replicate the issue for a while now.

I have had it fail on

 * a dependency on another app in the umbrella
 * another file in the same app

We can maybe rule out a few things here.

@myronmarston:

 * Are you using an umbrella app?
 * Could you provide the deps and applications list in your mix.exs?

**Comment by josevalim** [comment]
I don't think this is related to particular dependencies given all the different scenarios this is happening. In fact, I have fixed another bug in master that may have been the cause of such issues. Here is the rationale:

Elixir tracks modules being compiled by the compiler. For example, imagine these two files:

```elixir
# a.ex
defmodule A do
  B.__info__(:module)
end

# b.ex
defmodule B do
  A.__info__(:module)
end
```

Elixir knows that one file is defining B and waiting on A and another is defining A and waiting on B. That's a deadlock and the compiler will let you know.

However, there was a bug in the compiler where we were tracking only the current module and not the full module hierarchy. Here is an example:

```elixir
# a.ex
defmodule A do
  B.__info__(:module)
end

# b.ex
defmodule B do
  defmodule C do
    A.__info__(:module)
  end
end
```

Before Elixir would only see that module A is waiting on B and module C is waiting on A. We have fixed this so it is clear that both modules B and C are waiting on A, so we still detect the deadlock properly.

The reason why this can be the source of this bug is because of the following. We know that "lib/parse/datetime/parser.ex" is waiting for a Timex.DateTime struct definition. But if for some reason Timex.DateTime is waiting on something else and Elixir thinks that the dependency Timex.DateTime will define never be available (because for example, Timex.DateTime was waiting on B and a module C inside B was blocking), those failures would cascade and we would get a report on the wrong dependency.

For example, I found this bug because Ecto.Associations was complaining that Ecto.Query was not available. But the root cause was that Decimal library was not compiled and a module inside Ecto.Query (not Ecto.Query itself) was failing to compile due to the Decimal dependency. Ecto.Associations was unable to see Ecto.Query was being defined anywhere, leading to the wrong report.

Here is the commit with the fix: https://github.com/elixir-lang/elixir/commit/2a1d3d65303c96da56200d29a333310adfd8bace

This fix may not actually solve this issue, specially in @Gazler case where a module being defined in a dependency was not available, but the plan is to release a new RC this weekend or early next week and we will keep watching this closely.

Thanks everyone for reporting those bugs, it is extremely important!

**Comment by josevalim** [comment]
@Gazler said this definitely did not fix the issue, so we need to continue investigating. On the other hand, @Gazler also said that when this happens once it happens consistently.

**Comment by josevalim** [comment]
Fixed by #4777. 🎉

**Comment by myronmarston** [comment]
Thank you both, @josevalim and @Gazler!  I'm going to keep working on upgrading our app to 1.3 using HEAD of this repo (only changes meant for 1.3.0 final are being merged to master, right?) so I'll let you know if I see this again.
